doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,167 @@
+
+All reads to any object and all writes to user tables are assigned a timestamp by the Coordinator on a single thread.
+The logic for determining what timestamp to return must be altered to satisfy the following properties:
+
+- If it's a read, then it is larger than or equal to the `since` of all objects involved in the query (currently
+  satisfied).
+- If it's a write, then it is larger than or equal to the `upper` of the table (currently satisfied).
+- It is larger than or equal to the timestamp of the most recently completed query (not currently satisfied).
+    - If the query is a write and the most recently completed query was a read, then the timestamp returned must be
+      strictly larger than the previous timestamp.
+
+In order to achieve this, all timestamps returned must be larger than or equal to the global timestamp. Currently, we
+only look at the global timestamp for user table related queries, however we need to start looking at it for all
+queries.
+
+If some operation is given timestamp `ts`, then once that operation completes the global timestamp must be larger than
+or equal to `ts`. How this specific property is achieved for reads and writes is described in sections below.
+
+Additionally, the `global_timeline` should continue to increase every X seconds by X seconds to roughly track the system
+clock. As of writing this document, the current value for X is 1.
+
+The properties described in this section are sufficient to ensure Strict Serializability across one start of
+Materialize (i.e. not between restarts).
+
+- [X] All reads and writes occur at specific timestamps: satisfied by the Coordinator assigning timestamps to all
+  operations.
+- [X] These timestamps are non-decreasing in real-time order: all timestamps are assigned sequentially on a single
+  thread. Each timestamp is guaranteed to be larger than or equal to the timestamp of the most recently completed query.
+- [X] With strict increases for writes following reads: the global timestamp explicitly increases after all writes.
+- [X] Each timestamp is assigned to an event at some real-time moment within the event's bounds: For any two events e1
+  and e2, if e2 starts after e1 finished, then e2 will have a larger or equal timestamp.
+- [X] Each timestamp transition is made durable before any subsequent response is issued: writes wait for an `append`
+  command to be processed before returning a response to the client.
+    - NOTE: There is ongoing work to ensure that an `append` command doesn't return until it is durable.
+
+### Global Timestamp Recovery
+
+After a restart the Coordinator must reestablish the global timestamp to some value greater than or equal to the value
+of the previous global timestamp before the restart. This ensures that a read or write after the restart is assigned a
+timestamp greater than or equal to all reads and writes before the restart.
+
+TODO (The Simplest way is to make each timestamp durable but this is inefficient).
```

Comment:
Previous discussion can be found here: https://github.com/MaterializeInc/materialize/pull/12475#discussion_r875052606